### DISCUSSION TITLE
Extra yaml emitted when a function calls knitr::asis_output

### DISCUSSION THREAD

**Discussion by harrelfe** [quarto-dev/quarto-cli#4248] [Q&A] [answered]
I am writing a function that can be called several times within one code chunk.  It takes constructed HTML and wraps it in `knitr::asis_output()` so that I don't need to remember to put `results='asis'` in the chunk header.  The output is rendered correctly when running `Quarto`.  But I'm getting these two lines surrounding the rendered HTML:

````
::: {.cell-output-display}
::::
````

These extra lines are appearing verbatim in the finished `Quarto` HTML report.

:new: I think the constructed HTML that triggers the problem contains `<style>, <div>, <span>`.

Any ideas what is causing this and how to prevent it?  Thanks in advance.  The little function that calls `asis_output` given a string `x` of HTML appears below.

```
rendHTML <- function(x) {
  x <- paste0('\n', paste(x, collapse='\n'), '\n')
  if(length(getOption('knitr.in.progress'))) {
    return(knitr::asis_output(x))
    }
  print(htmltools::browsable(htmltools::HTML(x)))
}
```

See also https://stackoverflow.com/questions/75353004

**Comment by cscheid** [comment, marked as answer]
Those lines are not yaml. That's a markdown div, and it's generated (on purpose) if your cell doesn't contain `results='asis'`.

That markdown div should be getting converted to a div by Pandoc. Have you tried wrapping your raw HTML in

````
```{=html}
put-your-raw-html-here
```
````

instead of simply emitting the HTML output?

**Reply by harrelfe** [reply]
Thanks for the reply.  I need a method that doesn't require the user to create another chunk or to even know that html is being produced.  I set `options(prType='html')` at the top level when the functions produce html, and leave that off to have them produce plain text output.

Sounds like I need an option to turn off the div creation you mentioned.    It is odd that for a lot of the html I produce without `results='asis'` no divs are added.

**Reply by cscheid** [reply]
> Sounds like I need an option to turn off the div creation you mentioned.

Can you give a little more context around what you're trying to achieve? The issue is that we patch `knitr` to add that div, and it currently is not going to be very easy for you to get around that. The relevant code is here: https://github.com/quarto-dev/quarto-cli/blob/main/src/resources/rmd/patch.R#L96.

**Reply by harrelfe** [reply]
Here is the ultimate goal, as I'm updating the `Hmisc` and `rms` packages to make all their `html` methods seamlessly used inside and outside of `Quarto`, using `rendHTML` above.  Various functions are involved: descriptive statistics, data dictionary construction, LaTeX representation of fitted regression models (rendered by `Quarto` using e.g. `mathjax`), and regression fit summaries.  Here's a workflow:

````
```{r}
options(prType='html')   # defaults to 'plain'; there is also a 'latex' option, not considering this now
d <- data.frame( . . .)
contents(d)  # html table
describe(d)   # descriptive stats, with html includes micrographics
f <- ols(y ~ pol(x, 2))
f    # two nice html tables
latex(f)   # LaTeX algebraic representation of fitted model
```
````

If html is not in effect with `prType` the raw LaTeX markup is printed for `latex(f)`.

**Reply by cscheid** [reply]
Thank you, that helps. I don't quite understand the problem with the div we emit. What issues does that one cause in your output?

**Reply by harrelfe** [reply]
The big issue is that the divs appear verbatim in the final html report.  I'm going to try to narrow down why it doesn't appear with my simpler html-producing code.  :new: Please see next reply.

**Reply by harrelfe** [reply]
I captured some emitted HTML that triggers the problem.  You'll see 3 extraneous lines with divs that appear verbatim.  I don't get this when the emitted HTML code doesn't have all this style information.  This is with the latest `Quarto` preview release.

````
---
title: "Test"
format:
  html:
    self-contained: true
---


```{r}
x <- '
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" /> 
<style>
.earrows {color:silver;font-size:11px;}

fcap {
 font-family: Verdana;
 font-size: 12px;
 color: MidnightBlue
 }

smg {
 font-family: Verdana;
 font-size: 10px;
 color: &#808080;
}

hr.thinhr { margin-top: 0.15em; margin-bottom: 0.15em; }

span.xscript {
position: relative;
}
span.xscript sub {
position: absolute;
left: 0.1em;
bottom: -1ex;
}
</style>
 <title>stage Descriptives</title> <font color="MidnightBlue"><div align=center><span style="font-weight:bold">stage <br><br> 1  Variables   502  Observations</span></div></font> <hr class="thinhr"> <span style="font-weight:bold">stage</span>: Stage <style>
 .hmisctable468037 {
 border: none;
 font-size: 85%;
 }
 .hmisctable468037 td {
 text-align: center;
 padding: 0 1ex 0 1ex;
 }
 .hmisctable468037 th {
 color: MidnightBlue;
 text-align: center;
 padding: 0 1ex 0 1ex;
 font-weight: normal;
 }
 </style>
 <table class="hmisctable468037">
 <tr><th>n</th><th>missing</th><th>distinct</th><th>Info</th><th>Mean</th><th>Gmd</th></tr>
 <tr><td>502</td><td>0</td><td>2</td><td>0.733</td><td>3.424</td><td>0.4895</td></tr>
 </table>
 <pre style="font-size:85%;">
 Value          3     4
 Frequency    289   213
 Proportion 0.576 0.424
 </pre>
 <hr class="thinhr">
 '
knitr::asis_output(x)
```
````

**Reply by cscheid** [reply]
Can you try emitting

````
```{=html}
````

at the beginning of that block, and

````
```
````

at the end?

Like so:

`````
---
title: "Test"
format:
  html:
    self-contained: true
---


````{r}
x <- '
```{=html}
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" /> 
<style>
.earrows {color:silver;font-size:11px;}

fcap {
 font-family: Verdana;
 font-size: 12px;
 color: MidnightBlue
 }

smg {
 font-family: Verdana;
 font-size: 10px;
 color: &#808080;
}

hr.thinhr { margin-top: 0.15em; margin-bottom: 0.15em; }

span.xscript {
position: relative;
}
span.xscript sub {
position: absolute;
left: 0.1em;
bottom: -1ex;
}
</style>
 <title>stage Descriptives</title> <font color="MidnightBlue"><div align=center><span style="font-weight:bold">stage <br><br> 1  Variables   502  Observations</span></div></font> <hr class="thinhr"> <span style="font-weight:bold">stage</span>: Stage <style>
 .hmisctable468037 {
 border: none;
 font-size: 85%;
 }
 .hmisctable468037 td {
 text-align: center;
 padding: 0 1ex 0 1ex;
 }
 .hmisctable468037 th {
 color: MidnightBlue;
 text-align: center;
 padding: 0 1ex 0 1ex;
 font-weight: normal;
 }
 </style>
 <table class="hmisctable468037">
 <tr><th>n</th><th>missing</th><th>distinct</th><th>Info</th><th>Mean</th><th>Gmd</th></tr>
 <tr><td>502</td><td>0</td><td>2</td><td>0.733</td><td>3.424</td><td>0.4895</td></tr>
 </table>
 <pre style="font-size:85%;">
 Value          3     4
 Frequency    289   213
 Proportion 0.576 0.424
 </pre>
 <hr class="thinhr">
```'
knitr::asis_output(x)
````

`````

**Reply by cscheid** [reply]
The problem with the HTML you're emitting is that it confuses Pandoc's HTML reader. By wrapping it in

````
```{=html}
your-html-here
```
````

Pandoc will not attempt to parse your HTML as AST elements and get confused. Generally speaking, if you're in the business of emitting raw HTML, it pays off to use that syntax I suggested (unless you know precisely what you're doing and that it's safe).

**Reply by harrelfe** [reply]
That did it!  I'll try that now on the bigger test with dynamically emitted html.

**Reply by harrelfe** [reply]
The big test is working well.  You're the best!

**Reply by cscheid** [reply]
Glad to be of help!

**Comment by cderv** [comment]
I am adding context based on the discussion and solution I am seing above. 

You shared your function was 
````r
rendHTML <- function(x) {
  x <- paste0('\n', paste(x, collapse='\n'), '\n')
  if(length(getOption('knitr.in.progress'))) {
    return(knitr::asis_output(x))
    }
  print(htmltools::browsable(htmltools::HTML(x)))
}
````

I believe the existing HTML and R Markdown toolchain can do the right think for you without needing to add the raw attribute yourself and all like solution above. 

Doing the following should work ok in **knitr**
````r
rendHTML <- function(x) {
  x <- paste0('\n', paste(x, collapse='\n'), '\n')
  htmltools::browsable(htmltools::HTML(x))
}
````

When a code is run in knitting, the output will be processed by some specific `knit_print()` method (and not `print()`). More on that in https://cran.r-project.org/web/packages/knitr/vignettes/knit_print.html
When not running in **knitr**, the printing of your function result will be with `print()`

**htmltools** defines a `knit_print.html` method that will do 
* the HTML protecting for you adding the `` ```{=html} ``  
* output using `knitr::asis_output()`

So you should not need anything specific than using **htmltools** . 

Also FWIW, there is a `knitr::raw_html()` function that one can use to add the raw code attributes for Pandoc. 

Try that 
````markdown
---
title: raw HTML
format: html
---
    
```{r}
rendHTML <- function(x) {
  x <- paste0('\n', paste(x, collapse='\n'), '\n')
  htmltools::browsable(htmltools::HTML(x))
}
```


```{r}
rendHTML("<div>\ncontent\n</div>")
```
````

Hope it helps understand how this works
